docs/design/constraints.md

```diff
@@ -0,0 +1,177 @@
+       },
+       "constraintVS": {
+           "type": "vs-components",
+           "args": {
+                "components": [            //requires Xamarin of version 1.0.0.0 - version 5.0.0.0.
+                    {
+                        "name": "Xamarin",
+                        "minVersion": "1.0.0.0",
+                        "maxVersion": "5.0.0.0"
+                    }
+                ],
+                "whatever": true 
+           }
+       }
+  }
+```
+Constraint definition key is a unique name. The value is an object that has two mandatory keys:
+- `type`: string. Constraint type. All implemented constraints ideally should have unique types. The type will be used to identify constraint component (as alternative to using guids). Case-sensitive.
+- `args`: array or object. Constraint arguments. Arguments depends on constraint implementation. The JSON array or object will be passed to constraint component for evaluation as string. The constraint implementation should be able to parse the content.
+
+Constraint configuration will be read when `template.json` is read and extracted to template engine cache. 
+It will be available in `ITemplateInfo`, therefore for constraint evaluation `template.json` won't be read, and cached value will be used instead. 
+New property `Constraints` will be added to `ITemplateInfo` of type `IReadOnlyList<TemplateConstraintInfo>` containing template constraints information.
+
+`TemplateConstraintInfo` has the following properties:
+- `string Type`: constraint type
+- `string Args`: JSON string containing args read from `template.json`. Args to be parsed when evaluating the constraint by constraint itself.
+
+### Types uniqueness consideration
+At the moment 3rd party constraints won't be supported. 1st party developers should ensure unique naming.
+When 3rd party constraints are supported, the way to check if the constraint type is already used should be considered.
+
+
+## Constraint component
+
+Constraint implementation should implement `ITemplateConstraintFactory` interface that creates `ITemplateConstraint`:
+```csharp
+    public interface ITemplateConstraintFactory : IIdentifiedComponent
+    {
+        /// <summary>
+        /// Unique constraint type. Should match the definition in `template.json`.
+        /// </summary>
+        string Type { get; }
+
+        /// <summary>
+        /// Creates new <see cref="ITemplateConstraint"/> based on current <see cref="IEngineEnvironmentSettings"/>.
+        /// </summary>
+        public Task<ITemplateConstraint> CreateTemplateConstraintAsync(IEngineEnvironmentSettings environmentSettings);
+    }
+```
+When creating the constraint, it is assumed that constraint will be initialized and its context will be evaluated.
+Example: if constraint needs to know the project MSBuild properties it is assumed that project will be identified and MSBuild properties will be evaluated when the constraint is being created prior to first evaluation.
```

Comment:
I don't know that we can really do this at creation, since we won't have the args from the template.json until Evaluate is called.